Add `pkgs.lib.encodeGNUCommandLine`

Gabriella439 opened the pull request:
###### Motivation for this change

This adds a new utility to intelligently convert Nix records to
command line options to reduce boilerplate for simple use cases and to
also reduce the likelihood of malformed command lines

###### Things done

I ran `nix-instantiate --eval --strict lib/tests/misc.nix`

Gabriella439 committed:
Add `pkgs.lib.renderOptions`

This adds a new utility to intelligently convert Nix records to
command line options to reduce boilerplate for simple use cases and to
also reduce the likelihood of malformed command lines

edolstra commented:
I'm hesitant about this because it puts a highly specific policy of how command line arguments are encoded in the standard library. It's not at all clear how universally applicable this is. For example, `v = false` is encoded as an empty string, but maybe in some CLI tools you want it to be encoded as `--no-v`.

The name `renderOptions` is rather vague. At first I thought this had something to do with the NixOS module system. Maybe something like `encodeGNUCommandLine` is more descriptive.

Escaping shell arguments should be handled elsewhere since you could also pass arguments as an array. So escaping could be done separately (e.g. `concatMap escapeShellArg (renderOptions ...)`).

Gabriel439 commented:
@edolstra: I would be open to naming this in such a way that it does not imply that it is *the* only way to render command-line options and/or making the utility's behavior customizable (e.g. customizing how it handles lists or boolean-valued fields).  For example, conceptually this utility could take an additional record argument specifying how to handle each behavior (with default values so that the user can supply `{ }` for the common case):

```nix
encodeGNUCommandLine =
    { renderBool ? key: value: if value then " ${hyphenate key}" else ""
    , renderList ? key: values: ...
    , renderOption ? key: value: ...
    , ...
    }
    options: ...
```

Gabriella439 committed:
Rename `renderOptions` to `encodeGNUCommandLine`

... as suggested by @edolstra

Gabriella439 committed:
Make behavior of `encodeGNUCommandLine` customizable

... based on feedback from @edolstra

Gabriel439 commented:
@edolstra: Alright, I believe this is ready for re-review now.  I incorporated your feedback

roberth commented on lib/tests/misc.nix:
```diff
@@ -441,4 +441,21 @@ runTests {
+           qux = true;
+
+           v = true;
+         };
```
This should also have list examples and a `false` value.

Gabriella439 committed:
Use a more realistic example that exercises all encodings

... as suggested by @roberth

This also caught a bug in rendering lists, which this change also fixes

Gabriel439 commented:
@roberth: I believe I addressed your feedback and this is ready for another review

jtojnar commented:
Where would this be used seeing it returns string? We are moving towards `__structuredAttrs`, where things like `configureFlags` will be lists: https://github.com/NixOS/nixpkgs/pull/72074

Gabriel439 commented:
@jtojnar: A `systemd` service specification is one example of where you can use this:

```nix
{ config, lib, pkgs, ... }:

let
  cfg = config.services.example;

  options = { inherit (cfg) host port; };

in
  { options.services.example = {
      enable = lib.mkEnableOption "example";

      host = lib.mkOption { ... };

      port = lib.mkOption { ... };
    };

    config.systemd.services.example = lib.mkIf cfg.enable {
      wantedBy = [ "multi-user.target" ];

      script = ''
        ${pkgs}/bin/example${lib.renderOptions options}
      '';
    };
  }
```

roberth commented:
You could extract a function `toGNUCommandLine` from this: `encodeGNUCommandLine = a: escapeShellArgs (toGNUCommandLine a)`

`toGNUCommandLine` will then be usable with structured attrs.

This will require a preceding space as in `the-cmd ${encodeGNUCommandLine attrs}`, but that's good because it's consistent with `escapeShellArgs` and the shell itself.

jtojnar commented:
Well, the function returns a string. To be useful for structured attrs, it would have to return a list.

roberth commented:
> return a list

precisely

Gabriella439 committed:
Factor out a `toGNUCommandLine` utility

... as suggested by @roberth

roberth commented on lib/default.nix:
```diff
@@ -120,6 +121,7 @@ let
       isOptionType mkOptionType;
     inherit (asserts)
       assertMsg assertOneOf;
+    inherit (cli) encodeGNUCommandLine;
```
```suggestion
    inherit (cli) encodeGNUCommandLine toGNUCommandLine;
```

Gabriella439 committed:
Export toGNUCommandLine

... as suggested by @roberth

Co-Authored-By: Robert Hensing <roberth@users.noreply.github.com>

infinisil reviewed (approved):
I like it, this could be used to provide something like NixOS/rfcs#42 for CLI configs.

Profpatsch commented:
Just seeing this, it would have been good if it had followed the same interface as `lib/generators.nix`;
that is naming the options `mkX` instead of `renderX`. The `fn {} val` interface is the same, luckily.

Not entirely sure what the separation in `renderList/Option/Bool` is for, one argument with pattern match would suffice I think. As long as it’s new like this, can we still change the interface?

Profpatsch commented:
Oh, I’d also rename the functions to use `toX` for the string output (which is how generators.nix names its functions) and `toXList` for the list version.

Profpatsch commented:
Ah, no, the string is explicitely POSIX shell-formatted, so we should call it that. `toXShell`. See `escapeShellArg` in `strings.nix.`

Profpatsch commented:
Let’s also not expose it into the global namespace, similar to `generators`. `cli` is only three letters.

roberth commented:
@Profpatsch feel free to make a PR for the rename/move.

> Not entirely sure what the separation in renderList/Option/Bool is for

It's for overriding the convention, for example to allow `x = false` to be rendered as `--no-x`.

Re POSIX: it does seem to be GNU, at least according to GNU.

> GNU adds long options to these conventions. Long options consist of ‘--’ followed by a name made of alphanumeric characters and dashes.

-- https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html

I'd welcome a more generic term than GNU, but I don't know of any that implies long options.

Profpatsch commented:
I created a PR with my suggested breaking changes: https://github.com/NixOS/nixpkgs/pull/78337

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/january-2020-in-nixos/5771/1
